lib/licensee/projects/fs_project.rb

```diff
@@ -44,13 +44,18 @@ def files
         end
       end
 
-      # Retrieve a file's content from disk
+      # Retrieve a file's content from disk, enforcing encoding
       #
       # file - the file hash, with the :name key as the file's relative path
       #
       # Returns the file contents as a string
       def load_file(file)
-        File.read dir_path.join(file[:dir], file[:name])
+        content = File.read dir_path.join(file[:dir], file[:name])
+        content.force_encoding(ProjectFiles::ProjectFile::ENCODING)
+
+        return content if content.valid_encoding?
+
+        return content.encode(ProjectFiles::ProjectFile::ENCODING, ProjectFiles::ProjectFile::ENCODING_OPTIONS)
```

Comment:
Any idea why this is only needed in `fs_project` and not `git_project`? That makes me a bit suspicious (without having dug deeper). With the added fixture and without this change, does only the filesystem test fail (every fixture is tested twice, once in a fs project, once in a git project, IIRC).

Reply:
After some digging, I can clarify the difference. This may be more detail than you wanted.
- [`GitProject#load_file`](url) uses `Rugged::Blob.to_buffer` which calls down to a [C Binding](https://github.com/libgit2/rugged/blob/master/ext/rugged/rugged_blob.c#L496). 
- The [`rb_str_new`](https://github.com/libgit2/rugged/blob/master/ext/rugged/rugged_blob.c#L529) call in that method is a Ruby C method which converts a `char*` to a Ruby String with a default encoding of `ASCII-8BIT`. https://silverhammermba.github.io/emberb/c/#string. (Can't find official docs.)

The `FSProject#load_file` is using `File.read` which automatically assumes a default encoding of `UTF-8`.

So the same byte that's an invalid `UTF-8` character isn't an invalid `ASCII-8BIT` character and the `GitProject` tests actually don't fail.

```ruby
# UTF-8
f = File.read('spec/fixtures/readme-invalid-encoding/README.md')
=> "# Readme fixture\n\n ... \n\u0010\xFF\xFF"
f.encoding
=> #<Encoding:UTF-8>
f.valid_encoding?
=> false

# ASCII-8BIT
# Use invocation that specifies a length and forces ASCII-8BIT, emulating rb_str_new behavior
# https://ruby-doc.org/core-2.5.0/IO.html#method-i-read
f_ascii = File.read('spec/fixtures/readme-invalid-encoding/README.md', 1087)
=> "# Readme fixture\n\n ... \n\u0010\xFF\xFF"
f_ascii.encoding
=> #<Encoding:ASCII-8BIT>
f_ascii.valid_encoding?
=> true
```

Reply:
Thanks!